aion/test/controllers/room_controller_test.exs

```diff
@@ -1,26 +1,71 @@
 defmodule Aion.RoomControllerTest do
   use Aion.AuthConnCase
 
-  alias Aion.{Room, Category, RoomCategory}
-  @valid_attrs %{description: "some content", name: "some content"}
-  @invalid_attrs %{}
+  alias Aion.{Room, Question}
+  alias Aion.RoomChannel.Monitor
+
+  @room %Room{description: "Here come dat boi", name: "Boi"}
+  @question %Question{content: "Content"}
+  @player "Stephan"
 
   test "lists all entries on index", %{conn: conn} do
     conn = get conn, room_path(conn, :index)
     assert json_response(conn, 200)["data"] == []
   end
 
   test "shows chosen resource", %{conn: conn} do
-    room = Repo.insert! %Room{}
+    room = Repo.insert! @room
     conn = get conn, room_path(conn, :show, room)
-    assert json_response(conn, 200)["data"] == %{"id" => room.id,
+    assert json_response(conn, 200)["data"] == %{
+      "id" => room.id,
       "name" => room.name,
-      "description" => room.description}
+      "description" => room.description,
+    }
   end
 
   test "renders page not found when id is nonexistent", %{conn: conn} do
     assert_error_sent 404, fn ->
       get conn, room_path(conn, :show, -1)
     end
   end
+
+  test "GET /counts with no rooms in db", %{conn: conn} do
+    conn = get conn, "/api/counts"
+    assert json_response(conn, 200)["data"] == []
+  end
+
+  test "GET /counts with empty room", %{conn: conn} do
+    %Room{id: room_id} = room = Repo.insert! @room
+
+    Monitor.create(room_id, current_question: @question)
+
+    conn = get conn, "/api/counts"
+
+    assert json_response(conn, 200)["data"] == [
+      %{
+        "id" => room.id,
+        "name" => room.name,
+        "description" => room.description,
+        "player_count" => 0,
+      }
```

Comment:
I think it should be asserted to `@room.id` etc since currently it asserts to the values that are returned by the insert, not actually inserted object. 

Imagine abstract situation where somehow `Repo.insert` implementation changed and now it reverses the name before returning the value - the test would remain true although we expect the name to stay the same.

Reply:
I think the given situation is indeed _really abstract_. If `Repo.insert` changed it would actually be good to have tests fail rather than look for bugs inside the main logic. Imagine it changed in a different way and our tests would still be alright, but somehow the logic would be different (for example instead of `{:ok, %User}` we would get `{%User, :ok}` somewhere in the code. There are places in the code where we use things returned by this function, I think it's much easier to analyze the root of the problem in a unit test than in the middle of some nested GenServer architecture.

Also, how would you want to move that out to `@room.id`? To have an id you have to insert to db first, the actual id may differ depending on the environment and order in which the tests are run. 

NOTE: the db is not dropped in any way between the tests.

Reply:
Makes sense.